### PR TITLE
Port TML from .NET 6 to .NET 8

### PR DESCRIPTION
**Pull request by Mirsario**
.NET 7 and .NET 8 have brought plenty of performance-related improvements and upgrades, as well as some upgrades to C#. However, we've been choosing to stay on .NET 6 for quite a while now, since .NET 7 also happened to drop support of Windows 7, 8.0 and 8.1. Now, a workaround for that has been found, and this PR is here in hopes to bring all the aforementioned goodness without any downsides.

## Checklist
- [x] Port TerrariaNetCore.
- [x] Port tModLoader.
- [x] Set `DOTNET_EnableWriteXorExecute=0` envvar in start scripts if on win7/win8/win8.1.
- [x] Get `tests/TilemapPerformance` working again.
    - [x] Create some benchmarks comparing .NET 6 against .NET 8 in WorldGen.
- [x] Update CI to use .NET 8.
- [x] User feedback on performance is positive.
- [x] Almost all mods are confirmed to be working.
    - Infernum broke due to meddling with Windows-only libraries that got removed. Authors were notified. 
- [x] Update the FNA fork to compile for .NET 8, so that building mods doesn't log warnings.
- [x] Rebuild TerrariaHooks.
- [x] @Solxanich Mitigate problems of .NET updates changing executable paths, causing user-declared process-tied settings such as those of Nvidia Control Panel to reset.
    - ~~Use scripts to maintain a `dotnet/current/` folder through symlinks/hardlinks/junctions?~~
    - Just install directly into `dotnet/`?
- [x] ~~Replace or update `DotNetZip.Reduced`/`Ionic.Zip.Reduced`. It's throwing exceptions.~~ Done by `#4109`.
- [x] Update the pre-packaged runtime on windows (`dotnet-runtime-6.0.14-win-x64.zip`).
- [x] Resolve builds failing at large if-else chains with SDK >=8.0.200. They work in 8.0.100 and in Visual Studio (which ships custom SDK claiming to be 8.0.20X).
    - We've split the `Projectile.SetDefaults`  method on the `TerrariaNetCore` patchstep to avoid this.

### CODE REVIEW

File: test/TilemapPerformance/Program.cs [hunk 1]
```diff
@@ -37,10 +42,11 @@
 	new ILHook(hookMethod, il =>
```
Comment (Chicken-Bones) [inline test/TilemapPerformance/Program.cs]: These will need to be stored somewhere, lest they get undone on GC.

**Comment by Mirsario** [comment]
Worldgen benchmarks!
```
1.4.4:
	GenWorld:       64.3s   62.1s   60.0s
	SaveWorld:      2.6s    2.7s    2.6s
	LoadWorld:      1.5s    1.5s    1.5s

feature/dotnet8:
	GenWorld:       58.1s   54.6s   50.6s
	SaveWorld:      2.3s    2.3s    2.3s
	LoadWorld:      1.4s    1.4s    1.4s
```

**Comment by Solxanich** [comment]
Note: For the resulting 'Modder Impact' write-up: include mention of `you can prompt the modder to write setx DOTNET_EnableWriteXorExecute 0 in cmd, + a PC restart, it does the same thing as manually navigating to the envvar menu ` for Windows 7/8/8.1

**Review by Solxanich — APPROVED** [review]
Approved in concept. 
No outstanding design goals

**Pull request merged**